07-performance.Rmd

```diff
@@ -4,16 +4,24 @@ knit: "bookdown::preview_chapter"
 
 # Efficient optimization {#performance}
 
-[Donald Knuth](https://en.wikiquote.org/wiki/Donald_Knuth) is a famous American
-computer scientist who developed a number of the key algorithms that we use today.
-On the subject of optimisation he give this advice.
+[Donald Knuth](https://en.wikiquote.org/wiki/Donald_Knuth) is a legendary American
+computer scientist who developed a number of the key algorithms that we use today (see for 
+example `?Random`).
+On the subject of optimization he give this advice.
 
 > The real problem is that programmers have spent far too much time worrying about
 efficiency in the wrong places and at the wrong times; premature optimization is the
 root of all evil (or at least most of it) in programming.
 
 Knuth's point is that it is easy to undertake code optimisation inefficiently.
-For this reason this chapter is covered towards the latter half of the book. The previous chapters deliberately focussed on concepts, packages and functions
+When developing code the speed bottleneck is fluid and moves. This means that
+time spent optimizing code early in the developmental stage could be wasted. Even worse,
+there is a tradeoff between code speed and code readability; we've already made this trade-off
+once by using readable, (but slow) R compared with verbose C code!
+
+
+For this reason this chapter is covered towards the latter half of the book. 
+The previous chapters deliberately focussed on concepts, packages and functions
 to increase efficiency. These are (relatively) easy ways of saving time
 that, once implemented, will work for future projects.
 Code optimisation, by contrast, is an advanced topic that should only be tackled once 'low hanging fruit' for efficiency gains have been taken.
```

Comment:
We could add, as a kind of prerequisite, something along the lines of:

```
We assume in this chapter that you already have well-developed code that is mature conceptually has been tried and tested. Now you want to optimize this code, but not prematurely.
```

Verbose, but it's good to reinforce that point I think.